File: android/app/src/main/java/com/aurum/edge/engine/PerformanceMetrics.kt
package com.aurum.edge.engine

import com.aurum.edge.core.BacktestRecord
import com.aurum.edge.core.PaperTrade
import com.aurum.edge.core.SignalAction
import kotlin.math.sqrt

data class ClosedOutcome(
    val side: SignalAction,
    val entryAt: Long,
    val exitAt: Long,
    val pnlUsd: Double,
    val rMultiple: Double?,
)

data class PerformanceReport(
    val total: Int,
    val wins: Int,
    val losses: Int,
    val longCount: Int,
    val shortCount: Int,
    val winRatePct: Double?,
    val netPnlUsd: Double,
    val grossProfitUsd: Double,
    val grossLossUsd: Double,
    val averageWinUsd: Double?,
    val averageLossUsd: Double?,
    val profitFactor: Double?,
    /** Per-closed-trade Sharpe, risk-free rate zero, NOT annualized. Not daily Sharpe. */
    val tradeSharpe: Double?,
    val expectancyR: Double?,
    val averageDurationMillis: Long?,
    val longestWinningStreak: Int,
    val longestLosingStreak: Int,
    val maxDrawdownPct: Double?,
)

/** Descriptive statistics from settled outcomes only; no future return or synthetic observation. */
object PerformanceMetrics {
    fun fromBacktest(result: Backtester.Result): PerformanceReport = from(
        result.trades.map { ClosedOutcome(it.side, it.entryTime, it.exitTime, it.pnlUsd, it.rMultiple) },
        result.initialBalance,
    )

    fun fromStoredReport(record: BacktestRecord): PerformanceReport? {
        // Older versions stored only the last 80 trades; do not present partial metrics as full.
        if (record.trades.size != record.wins + record.losses) return null
        return from(record.trades.map { ClosedOutcome(
            if (it.side == "BUY") SignalAction.BUY else SignalAction.SELL,
            it.entryTime, it.exitTime, it.pnlUsd, it.rMultiple,
        ) }, record.initialBalance)
    }

    fun fromPaper(trades: List<PaperTrade>, startingBalance: Double): PerformanceReport = from(
        trades.filter { !it.isOpen && it.pnlUsd != null && it.closedAt != null }.map {
            ClosedOutcome(it.action, it.openedAt, it.closedAt!!, it.pnlUsd!!, it.rMultiple)
        }, startingBalance,
    )

    fun from(outcomes: List<ClosedOutcome>, startingBalance: Double): PerformanceReport {
        require(startingBalance.isFinite() && startingBalance > 0) { "موجودی اولیه باید مثبت باشد" }
        val closed = outcomes.filter { it.pnlUsd.isFinite() }.sortedBy { it.exitAt }
        val wins = closed.filter { it.pnlUsd > 0 }
        val losses = closed.filter { it.pnlUsd <= 0 } // breakeven counts as non-win
        val grossWin = wins.sumOf { it.pnlUsd }
        val grossLoss = -losses.sumOf { it.pnlUsd }
        val durations = closed.mapNotNull { (it.exitAt - it.entryAt).takeIf { ms -> ms > 0 } }
        val rValues = closed.mapNotNull { it.rMultiple?.takeIf { value -> value.isFinite() } }
        var bestWin = 0
        var bestLoss = 0
        var consecutiveWins = 0
        var consecutiveLosses = 0
        var equity = startingBalance
        var peak = equity
        var drawdown = 0.0
        val returns = mutableListOf<Double>()
        for (outcome in closed) {
            // Zero/negative equity makes a percentage return undefined, not zero.
            if (equity > 0) returns += outcome.pnlUsd / equity
            equity += outcome.pnlUsd
            peak = maxOf(peak, equity)
            drawdown = maxOf(drawdown, (peak - equity) / peak * 100)
            if (outcome.pnlUsd > 0) {
                consecutiveWins++
                consecutiveLosses = 0
            } else {
                consecutiveLosses++
                consecutiveWins = 0
            }
            bestWin = maxOf(bestWin, consecutiveWins)
            bestLoss = maxOf(bestLoss, consecutiveLosses)
        }
        val mean = returns.takeIf { it.size >= 2 }?.average()
        val variance = mean?.let { avg -> returns.sumOf { (it - avg) * (it - avg) } / (returns.size - 1) }
        val sharpe = if (returns.size == closed.size && mean != null && variance != null && variance > 0) mean / sqrt(variance) else null
        return PerformanceReport(
            total = closed.size,
            wins = wins.size, losses = losses.size,
            longCount = closed.count { it.side == SignalAction.BUY },
            shortCount = closed.count { it.side == SignalAction.SELL },
            winRatePct = if (closed.isEmpty()) null else wins.size * 100.0 / closed.size,
            netPnlUsd = closed.sumOf { it.pnlUsd },
            grossProfitUsd = grossWin, grossLossUsd = grossLoss,
            averageWinUsd = wins.takeIf { it.isNotEmpty() }?.map { it.pnlUsd }?.average(),
            averageLossUsd = losses.takeIf { it.isNotEmpty() }?.map { it.pnlUsd }?.average(),
            profitFactor = if (grossLoss > 0) grossWin / grossLoss else null,
            tradeSharpe = sharpe,
            expectancyR = rValues.takeIf { it.isNotEmpty() }?.average(),
            averageDurationMillis = durations.takeIf { it.isNotEmpty() }?.average()?.toLong(),
            longestWinningStreak = bestWin, longestLosingStreak = bestLoss,
            maxDrawdownPct = if (closed.isEmpty()) null else drawdown,
        )
    }
}

File: backend/app/services/performance_metrics.py
"""Descriptive report from actual settled backtest/journal trade records only.

Sharpe is per-trade (sample standard deviation, zero risk-free rate), NOT annualized
and NOT a daily Sharpe. No annualization without an observed daily equity series.
"""
from __future__ import annotations

import math
from datetime import datetime
from statistics import mean, stdev
from typing import Any


def summarize_trades(trades: list[dict[str, Any]], initial_balance: float) -> dict[str, Any]:
    if not math.isfinite(initial_balance) or initial_balance <= 0:
        raise ValueError("initial_balance must be positive")
    settled = sorted((t for t in trades if t.get("pnl") is not None and math.isfinite(t["pnl"])),
                     key=lambda t: t.get("exit_time", ""))
    winners = [t for t in settled if t["pnl"] > 0]
    losers = [t for t in settled if t["pnl"] <= 0]
    gross_profit = sum(t["pnl"] for t in winners)
    gross_loss = abs(sum(t["pnl"] for t in losers))
    durations: list[float] = []
    returns: list[float] = []
    r_multiples: list[float] = []
    equity = initial_balance
    peak = equity
    max_dd = 0.0
    win_streak = loss_streak = best_win = best_loss = 0
    for trade in settled:
        opened = trade.get("entry_time")
        closed = trade.get("exit_time")
        if opened and closed:
            try:
                seconds = (datetime.fromisoformat(closed) - datetime.fromisoformat(opened)).total_seconds()
                if seconds > 0:
                    durations.append(seconds)
            except ValueError:
                pass
        r = trade.get("r_multiple")
        if isinstance(r, (int, float)) and math.isfinite(r):
            r_multiples.append(r)
        if equity > 0:
            returns.append(trade["pnl"] / equity)
        equity += trade["pnl"]
        peak = max(peak, equity)
        max_dd = max(max_dd, (peak - equity) / peak * 100)
        if trade["pnl"] > 0:
            win_streak, loss_streak = win_streak + 1, 0
        else:
            win_streak, loss_streak = 0, loss_streak + 1
        best_win, best_loss = max(best_win, win_streak), max(best_loss, loss_streak)
    sharpe = mean(returns) / stdev(returns) if len(returns) == len(settled) and len(returns) >= 2 and stdev(returns) > 0 else None
    return {
        "total": len(settled), "wins": len(winners), "losses": len(losers),
        "long_count": sum(t.get("side") == "BUY" for t in settled),
        "short_count": sum(t.get("side") == "SELL" for t in settled),
        "win_rate_pct": round(len(winners) / len(settled) * 100, 2) if settled else None,
        "net_pnl_usd": round(sum(t["pnl"] for t in settled), 2),
        "gross_profit_usd": round(gross_profit, 2), "gross_loss_usd": round(gross_loss, 2),
        "average_win_usd": round(mean(t["pnl"] for t in winners), 2) if winners else None,
        "average_loss_usd": round(mean(t["pnl"] for t in losers), 2) if losers else None,
        "profit_factor": round(gross_profit / gross_loss, 3) if gross_loss else None,
        "sharpe_per_trade": round(sharpe, 3) if sharpe is not None else None,
        "expectancy_r": round(mean(r_multiples), 3) if r_multiples else None,
        "average_duration_seconds": round(mean(durations)) if durations else None,
        "longest_winning_streak": best_win, "longest_losing_streak": best_loss,
        "max_drawdown_pct": round(max_dd, 2) if settled else None,
        "fees_usd": round(sum(t.get("fees", 0.0) for t in settled), 2),
        "method": "per-trade Sharpe; sample stdev; zero risk-free; not annualized",
    }
